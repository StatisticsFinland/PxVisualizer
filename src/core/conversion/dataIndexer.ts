import { IQueryVisualizationResponse } from "../types";
import { EVariableType, IVariableMeta, IVariableValueMeta, TMultiLanguageString } from "../types/queryVisualizationResponse";
import { TVariableSelections } from "../types/variableSelections";
import { IDataCell, IDataSeries } from "../types/view";
import { cartesianProduct } from "./utilityFunctions";
import { isEqual } from "lodash";

export class DataIndexer {
    public dataIndex: number;
    public dataLength: number;

    private responseObj: IQueryVisualizationResponse;
    private coordinates: number[][];
    private lastIndices: number[];
    private indices: number[];
    private variableOrder: number[];
    private lastCoordinateIndex: number;
    private reverseCumulativeProducts: number[];
    private contentVariableIndex: number;
    private timeVariableIndex: number;
    private targetMap: IVariableMeta[];

    constructor(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections) {
        this.responseObj = responseObj;
        const completeMap: IVariableMeta[] = responseObj.metaData;
        const variableSizes: number[] = completeMap.map(v => v.values.length);
        this.targetMap = this.getTargetMap(responseObj, selectedValueCodes);
        this.coordinates = Array.from({ length: this.targetMap.length });
        this.variableOrder = this.getVariableOrder(completeMap, this.targetMap);
        for (let targetIndex = 0; targetIndex < this.targetMap.length; targetIndex++) {
            const variableIndex: number = this.variableOrder[targetIndex];
            const targetVariableCode = this.targetMap[targetIndex].code;
            const values: IVariableValueMeta[] | undefined = completeMap.find(v => v.code === targetVariableCode)?.values;
            if (!values) throw new Error("Provided variable code can not be found from the metadata");
            const targetVariable: IVariableMeta | undefined = this.targetMap.find(v => v.code === targetVariableCode);
            if (!targetVariable) throw new Error("Provided variable code can not be found from the target map");
            this.coordinates[variableIndex] = Array.from({ length: targetVariable.values.length });
            for (let mapIndex = 0; mapIndex < targetVariable.values.length; mapIndex++) {
                for (let valueIndex = 0; valueIndex < values.length; valueIndex++) {
                    if (values[valueIndex] === targetVariable.values[mapIndex]) {
                        this.coordinates[variableIndex][mapIndex] = valueIndex;
                    }
                }
            }
        }
        this.contentVariableIndex = this.variableOrder.indexOf(this.targetMap.findIndex(v => v.type === EVariableType.Content));
        this.timeVariableIndex = this.variableOrder.indexOf(this.targetMap.findIndex(v => v.type === EVariableType.Time));
        this.indices = Array.from({ length: completeMap.length }, () => 0);
        this.lastIndices = this.targetMap.map(v => v.values.length - 1);
        this.lastCoordinateIndex = this.coordinates.length - 1;
        this.reverseCumulativeProducts = this.generateRCP(variableSizes);
        this.dataLength = this.targetMap.map(v => v.values.length).reduce((acc, val) => acc * val, 1);
        this.dataIndex = 0;
        this.setCurrentIndex();
    }

    public getViewSeries(): IDataSeries[] {
        const viewSeries: IDataSeries[] = this.generateViewSeries();
        do {
            const dataCell = this.createDataCell();
            const rowNames: TMultiLanguageString[] = this.getRowNames();
            this.getRowNameGroup(viewSeries, rowNames).series.push(dataCell);
        } while (this.next());

        return viewSeries;
    }

    generateViewSeries(): IDataSeries[] {
        const rowValues: IVariableValueMeta[][] = this.targetMap.filter(v =>
            this.responseObj.rowVariableCodes.includes(v.code) &&
            (v.values.length > 1)).map(v => v.values);
        const multiSelectableValues: IVariableValueMeta[][] = this.targetMap.filter(v =>
            this.responseObj.visualizationSettings.multiselectableVariableCode === v.code).map(v => v.values);
        const nameGroupValues: IVariableValueMeta[][] = multiSelectableValues.concat(rowValues);
        const cartesianRowValues: IVariableValueMeta[][] = cartesianProduct(nameGroupValues);
        const viewSeries: IDataSeries[] = cartesianRowValues.reduce((acc, row) => {
            const nameGroup: TMultiLanguageString[] = row.map(v => v.name);
            acc.push({ rowNameGroup: nameGroup, series: [] });
            return acc;
        }, [] as IDataSeries[]);
        return viewSeries;
    }

    next(): boolean {
        for (let i = this.lastCoordinateIndex; i >= 0; i--) {
            const variableIndex: number = this.variableOrder[i];
            if (this.indices[variableIndex] < this.lastIndices[i]) {
                this.indices[variableIndex]++;
                this.setCurrentIndex();
                return true;
            }
            else {
                this.indices[variableIndex] = 0;
            }
        }
        return false;
    }

    createDataCell(): IDataCell {
        const timeVal: IVariableValueMeta | undefined = this.getCurrentTimeValue();
        const dataCell: IDataCell = {
            value: this.responseObj.data[this.dataIndex],
            precision: this.getCurrentContentValue().contentComponent?.numberOfDecimals ?? 0,
            preliminary: timeVal ? Object.values(timeVal.name)[0].trim().endsWith('*') : false
        };
        if (!dataCell.value) dataCell.missingCode = this.responseObj.missingDataInfo[this.dataIndex];
        return dataCell;
    }

    getRowNames(): TMultiLanguageString[] {
        let rowNames: TMultiLanguageString[] = [];
        for (let i = 0; i < this.indices.length; i++) {
            const variable: IVariableMeta = this.targetMap[i];
            const value: IVariableValueMeta = variable.values[this.indices[i]];
            if ((this.responseObj.rowVariableCodes.includes(variable.code) ||
                this.responseObj.visualizationSettings.multiselectableVariableCode === variable.code) &&
                variable.values.length > 1) {
                rowNames.push(value.name);
            }
        }
        return rowNames;
    }

    getRowNameGroup(viewSeries: IDataSeries[], rowNames: TMultiLanguageString[]): IDataSeries {
        const match: IDataSeries | undefined = viewSeries.find(vs => isEqual(vs.rowNameGroup, rowNames));
        if (match) {
            return match;
        }
        throw new Error("Provided row name group can not be found from the view series");
    }

    getTargetMap(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections): IVariableMeta[] {
        const targetMap: IVariableMeta[] = responseObj.metaData.map((variable) => {
            if (responseObj.selectableVariableCodes.includes(variable.code)) {
                const values: IVariableValueMeta[] = variable.values.filter((value) =>
                    selectedValueCodes[variable.code].includes(value.code)
                );
                if (values.length === 0) {
                    throw new Error("Provided selected value code can not be found from the metadata");
                }
                return { ...variable, values };
            }
            return variable;
        });
        return targetMap;
    }

    getVariableOrder(completeMap: IVariableMeta[], targetMap: IVariableMeta[]): number[] {
        const sourceCodes: string[] = completeMap.map(v => v.code);
        return targetMap.map(v => sourceCodes.indexOf(v.code));
    }

    generateRCP(variableSizes: number[]): number[] {
        let numOfDims: number = variableSizes.length;
        const cnt: number[] = Array.from({ length: numOfDims });
        let cumulativeMultiplier: number = 1;
        for (let i = numOfDims - 1; i >= 0; i--) {
            cnt[i] = cumulativeMultiplier;
            cumulativeMultiplier *= variableSizes[i];
        }
        return cnt;
    }

    setCurrentIndex() {
        this.dataIndex = 0;
        for (let i = 0; i <= this.lastCoordinateIndex; i++) {
            const variableIndex: number = this.variableOrder[i];
            this.dataIndex += this.reverseCumulativeProducts[variableIndex] * this.coordinates[variableIndex][this.indices[variableIndex]];
        }
    }

    getCurrentContentValue(): IVariableValueMeta {
        return this.targetMap[this.contentVariableIndex].values[this.indices[this.contentVariableIndex]];
    }

    getCurrentTimeValue(): IVariableValueMeta | undefined {
        if (this.timeVariableIndex == -1) {
            return undefined;
        }
        return this.targetMap[this.timeVariableIndex].values[this.indices[this.timeVariableIndex]];
    }
}