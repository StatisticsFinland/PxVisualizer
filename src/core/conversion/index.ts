export { convertPxGraphDataToChartOptions } from './pxGrafDataConverter';
export { extractSelectableVariableValues } from './helpers';
export { convertPxGrafResponseToView } from './viewUtils';
export { TTranslationPackage, AddTranslations } from './translations';