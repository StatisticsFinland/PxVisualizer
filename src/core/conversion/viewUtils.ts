import { IQueryVisualizationResponse } from "../types";
import { EVariableType, EVisualizationType, IContentComponent, IVariableMeta, IVariableValueMeta, TMultiLanguageString } from "../types/queryVisualizationResponse";
import { ESeriesType, IDataCell, IDataSeries, IUnitInfo, TValueSelectionAmounts, View } from "../types/view";
import { ASCENDING, ASCENDING_SORTING_FUNC, DESCENDING, DESCENDING_SORTING_FUNC, NO_SORTING, SUM, REVERSED, sortViewBasedOnSeries, sortViewBasedOnSeriesRelative, sortViewBasedOnSum, reverseViewOrder } from "./viewSorting";
import { cartesianProduct, onlyUnique } from "./utilityFunctions";
import { TVariableSelections } from "../types/variableSelections";
import Translations from "./translations";

export function convertPxGrafResponseToView(
    responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections
): View {
    const view: View = convert(responseObj, selectedValueCodes, getValueSelectAmounts(selectedValueCodes));

    const isRelativeChart =
        view.visualizationSettings?.visualizationType === EVisualizationType.PercentHorizontalBarChart
        || view.visualizationSettings?.visualizationType === EVisualizationType.PercentVerticalBarChart;

    switch (responseObj.visualizationSettings?.sorting) {
        case null: return view;
        case undefined: return view;
        case NO_SORTING: return view;
        case REVERSED: return reverseViewOrder(view);
        case SUM: return sortViewBasedOnSum(view, DESCENDING_SORTING_FUNC);
        case ASCENDING: return sortViewBasedOnSeries(view, 0, ASCENDING_SORTING_FUNC);
        case DESCENDING: return sortViewBasedOnSeries(view, 0, DESCENDING_SORTING_FUNC);
        default:
            if (isRelativeChart) return sortViewBasedOnSeriesRelative(view, convertToRelative(view), getSortingIndex(responseObj), DESCENDING_SORTING_FUNC);
            else return sortViewBasedOnSeries(view, getSortingIndex(responseObj), DESCENDING_SORTING_FUNC);
    };
}

export function convertToRelative(input: View): View {
    const newSeries = input.series.map(s => ({
        ...s,
        series: s.series.map((dataCell, index) => {
            const sum = input.series.reduce((sum, series) => sum + (series.series[index].value ?? 0), 0);
            const value = (!sum || sum === 0) ? null : Math.round((100 * (dataCell.value ?? 0) / sum) * 10) / 10;
            const newCell: IDataCell = { value, precision: 1, preliminary: dataCell.preliminary };
            if (value === null) newCell.missingCode = dataCell.missingCode;
            return newCell;
        }),
    }));

    const convertUnit = (unit: TMultiLanguageString) => {
        return Object.keys(unit).reduce((acc: TMultiLanguageString, lang: string) => {
            acc[lang] = '%';
            return acc;
        }, {})
    };

    return {...input, units: input.units.map(u => ({name: u.name, unit: convertUnit(u.unit) })) , series: newSeries};
}

function getValueSelectAmounts(selectedValueCodes: TVariableSelections): TValueSelectionAmounts {
    const result: TValueSelectionAmounts = {};
    Object.entries(selectedValueCodes).forEach(([key, value]) => result[key] = value.length);
    return result;
}

function convert(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections, selectedValueAmounts: TValueSelectionAmounts): View {
    const { metaData, selectableVariableCodes } = responseObj;
    const selectableVariables = getSelectableVariables(metaData, selectableVariableCodes);
    const contentVar = getContentVariable(metaData);
    const unsortedSeries = buildSeries(responseObj, selectedValueCodes);

    return {
        header: responseObj.header,
        tableReferenceName: responseObj.tableReference.name,
        subheaderValues: getSubheaderValues(selectableVariables, selectedValueCodes, selectedValueAmounts),
        units: getUnitInformation(contentVar, selectedValueCodes),
        sources: getContentProperty(contentVar, selectedValueCodes, (cc) => cc?.source ?? Translations.empty),
        columnNameGroups: unsortedSeries.columnNameGroups,
        series: unsortedSeries.series,
        colVarNames: getVariableNames(responseObj.columnVariableCodes, responseObj.metaData),
        rowVarNames: getVariableNames(responseObj.rowVariableCodes, responseObj.metaData),
        selectableVarNames: getVariableNames(responseObj.selectableVariableCodes, responseObj.metaData),
        visualizationSettings: responseObj.visualizationSettings,
        seriesType: getSeriesType(responseObj.columnVariableCodes, responseObj.metaData)
    };
}

export function buildSeries(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections): { columnNameGroups: TMultiLanguageString[][], series: IDataSeries[] } {
    const valuesInView: Set<string> = new Set(getValuesInView(responseObj, selectedValueCodes).map(v => v.code));
    const rowVarValues: IVariableValueMeta[][] = getVariableValues(responseObj, responseObj.rowVariableCodes);
    const columnVarValues: IVariableValueMeta[][] = getVariableValues(responseObj, responseObj.columnVariableCodes);
    const selectableValues: IVariableValueMeta[][] = getVariableValues(responseObj, responseObj.selectableVariableCodes);
    const cartesianColumnVarValues: IVariableValueMeta[][] = cartesianProduct(columnVarValues);
    const cartesianRowVarValues: IVariableValueMeta[][] = cartesianProduct(rowVarValues);
    const cartesianSelectableVarValues: IVariableValueMeta[][] = cartesianProduct(selectableValues);
    const viewPrecision: number | null = getViewPrecision(responseObj, selectedValueCodes);
    const multiselectValues: IVariableValueMeta[] = responseObj.metaData.find(v => v.code === responseObj.visualizationSettings.multiselectableVariableCode)?.values ?? [];

    let dataIndex: number = 0;
    const viewSeries: IDataSeries[] = [];

    cartesianSelectableVarValues.forEach(selectableVarValueGroup => {
        const multiselects = selectableVarValueGroup.filter(v => multiselectValues.includes(v));
        cartesianRowVarValues.forEach(rowVarValueGroup => {
            const preliminaryRow: boolean = rowVarValueGroup.some(v => Object.values(v.name)[0].trim().endsWith('*'));
            const rowPrecision: number | null = viewPrecision ?? rowVarValueGroup.find(v => v.contentComponent)?.contentComponent?.numberOfDecimals ?? null;
            const rowSeries: IDataCell[] = [];

            cartesianColumnVarValues.forEach(colVarValueGroup => {
                const combinedValues: IVariableValueMeta[] = [...rowVarValueGroup, ...colVarValueGroup, ...selectableVarValueGroup];
                const dataCell: IDataCell | null = createDataCell(
                    combinedValues,
                    valuesInView,
                    responseObj,
                    dataIndex,
                    rowPrecision,
                    preliminaryRow,
                    colVarValueGroup);
                if (dataCell) {
                    rowSeries.push(dataCell);
                }
                dataIndex++;
            });

            if (rowSeries.length > 0) {
                viewSeries.push({
                    rowNameGroup: [...multiselects.map(value => value.name), ...rowVarValueGroup.map(value => value.name)],
                    series: rowSeries
                });
            }
        });
    });

    return {
        columnNameGroups: cartesianColumnVarValues.map(columnVarValueGroup => columnVarValueGroup.map(value => value.name)),
        series: viewSeries
    };
}

function createDataCell(
    combinedValues: IVariableValueMeta[],
    valuesInView: Set<string>,
    responseObj: IQueryVisualizationResponse,
    dataIndex: number,
    rowPrecision: number | null,
    preliminaryRow: boolean,
    colVarValueGroup: IVariableValueMeta[]
): IDataCell | null {
    if (combinedValues.every(value => valuesInView.has(value.code))) {
        const dataCell: IDataCell = {
            value: responseObj.data[dataIndex],
            precision: rowPrecision ?? colVarValueGroup.find(v => v.contentComponent)?.contentComponent?.numberOfDecimals ?? 0,
            preliminary: preliminaryRow || colVarValueGroup.some(v => Object.values(v.name)[0].trim().endsWith('*'))
        };

        if (!dataCell.value) dataCell.missingCode = responseObj.missingDataInfo[dataIndex];
        return dataCell;
    }
    else {
        return null;
    }
}

function getValuesInView(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections): IVariableValueMeta[] {
    return responseObj.metaData.flatMap(variable => {
        if (responseObj.selectableVariableCodes.includes(variable.code) ||
            Object.keys(selectedValueCodes).includes(variable.code)) {
            const values: IVariableValueMeta[] = variable.values.filter(value => selectedValueCodes[variable.code].includes(value.code));
            if (values.length === 0) throw new Error("Provided selected value code can not be found from the metadata");
            return values;
        }
        else {
            return variable.values;
        }
    });
}

function getVariableValues(responseObj: IQueryVisualizationResponse, variableCodes: string[]): IVariableValueMeta[][] {
    return responseObj.metaData
        .filter(vm => variableCodes.includes(vm.code))
        .filter(vm => vm.values.length > 1)
        .map(vm => vm.values);
}

function getViewPrecision(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections): number | null {
    const contentVar = getContentVariable(responseObj.metaData);

    // Only one content variable value
    if (contentVar.values.length === 1) {
        return contentVar.values[0].contentComponent?.numberOfDecimals ?? null

    // Content variable in selection
    } else if (responseObj.selectableVariableCodes.includes(contentVar.code) && responseObj.visualizationSettings.multiselectableVariableCode !== contentVar.code) {
        return contentVar.values.find(value => selectedValueCodes[contentVar.code].includes(value.code))?.contentComponent?.numberOfDecimals ?? null;

    } else {
        return null;
    }
}

function getVariableNames(varCodes: string[], meta: IVariableMeta[]): TMultiLanguageString[] {
    return varCodes.reduce((acc: TMultiLanguageString[], code: string) => {
        const name = meta.find(v => v.code === code)?.name;
        if (name) acc.push(name);
        return acc;
    }, [])
}

function getSelectableVariables(metaData: IVariableMeta[], selectableVariableCodes: string[]): IVariableMeta[] {
    return metaData.filter(varMeta => selectableVariableCodes.includes(varMeta.code));
}

function getSortingIndex(responseObj: IQueryVisualizationResponse): number {
    const sortingVar = responseObj.metaData.find(v => v.code == responseObj.rowVariableCodes[0]);
    const valIndex = sortingVar?.values.findIndex(v => v.code === responseObj.visualizationSettings?.sorting);
    return (!valIndex || valIndex < 0) ? 0 : valIndex;
}

function getContentVariable(varMeta: IVariableMeta[]): IVariableMeta {
    const contentVar = varMeta.find(v => v.type === EVariableType.Content);
    if (contentVar) return contentVar;
    else throw new Error('Content variable is not defined');
}

function getSubheaderValues(selectableVariables: IVariableMeta[], selectedValueCodes: TVariableSelections, selectedValueAmounts: TValueSelectionAmounts): TMultiLanguageString[] {
    const subheaderValues: TMultiLanguageString[] = [];
    selectableVariables.forEach(selectableVariable => {
        // We only want to show the value name if there is exactly one selected.
        if (selectedValueAmounts[selectableVariable.code] == 1) {
            // Find the value whose code matches selected value code for this variable and add it to subheaderValues
            subheaderValues.push(selectableVariable.values.find(v => v.code === selectedValueCodes[selectableVariable.code][0])?.name ?? {})
        }
    });
    return subheaderValues;
}

function getUnitInformation(contentVar: IVariableMeta, selectedValueCodes: TVariableSelections): IUnitInfo[] {
    const values = contentVar.code in selectedValueCodes
        ? contentVar.values.filter(v => selectedValueCodes[contentVar.code].includes(v.code))
        : contentVar.values

    return values.reduce((accum: IUnitInfo[], valMeta: IVariableValueMeta) => {
        if(valMeta.contentComponent) accum.push({
            name: valMeta.name,
            unit: valMeta.contentComponent.unit
        });
        return accum;
    }, []);
}

function getContentProperty(
    contentVar: IVariableMeta,
    selectedValueCodes: TVariableSelections,
    extractorFunc: (cc: IContentComponent | null) => TMultiLanguageString
): TMultiLanguageString[] {
    if (contentVar.code in selectedValueCodes) {
        return contentVar.values
            .filter(v => selectedValueCodes[contentVar.code].includes(v.code))
            .map(cvv => extractorFunc(cvv.contentComponent))
            .filter(onlyUnique);
    } else {
        return contentVar.values.map(cvv => extractorFunc(cvv.contentComponent)).filter(onlyUnique);
    }
}

function getSeriesType (varCodes: string[], meta: IVariableMeta[]) {
    if (varCodes.length > 1 || varCodes.length === 0) {
        return ESeriesType.Nominal;
    }

    const metaDataType = meta.find(v => v.code === varCodes[0])?.type;

    switch (metaDataType) {
        case EVariableType.Time:
            return ESeriesType.Time;
        case EVariableType.Ordinal:
            return ESeriesType.Ordinal;
        default:
            return ESeriesType.Nominal;
    }
}