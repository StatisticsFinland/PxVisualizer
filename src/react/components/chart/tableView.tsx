import React, { useMemo } from "react";
import { View } from "../../../core/types/view";
import { v4 as uuidv4 } from 'uuid';
import { renderHtmlTable } from "../../../core/tables/htmlTable";

export interface ITableViewProps {
    view: View;
    locale: string;
    showTitles?: boolean;
    showUnits?: boolean;
    showSources?: boolean;
    footnote?: string;
}

export const TableView: React.FC<ITableViewProps> = ({ view, locale, footnote, showTitles = true, showUnits = true, showSources = true }) => {
    const uuid = useMemo(() => uuidv4(), [view, locale]);

    React.useEffect(() => {
        document.getElementById(uuid)?.replaceChildren();
        renderHtmlTable(view, locale, showTitles, showUnits, showSources, uuid, footnote);
    }, [view, locale, showTitles, showUnits, showSources, footnote]);

    return <div className={'tableChart'} id={uuid} />;
}