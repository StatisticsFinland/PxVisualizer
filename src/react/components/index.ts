export * from "./chart";
export * from "./ErrorBoundary";
export * from "./globalStyle";
export * from "./icon";
export * from "./burgerMenu";
