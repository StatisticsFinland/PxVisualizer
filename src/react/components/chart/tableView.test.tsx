import { render } from "@testing-library/react";
import { TableView } from "./tableView";
import { convertPxGrafResponseToView } from "../../../core/conversion/viewUtils";
import React from "react";
import { TABLE_WITH_ROW_AND_COLUMN_VARIABLES } from "../../../core/conversion/TestFixtures/tableChart";

jest.mock('uuid', () => ({
    ...jest.requireActual('uuid'),
    v4: () => 'foobar'
}));

describe('TableView render tests', () => {
    it('Should render correctly', () => {
        const { asFragment } = render(<TableView view={convertPxGrafResponseToView(TABLE_WITH_ROW_AND_COLUMN_VARIABLES.pxGraphData, TABLE_WITH_ROW_AND_COLUMN_VARIABLES.selectedVariableCodes)} locale={'fi'} />);
        expect(asFragment()).toMatchSnapshot();
    });
});