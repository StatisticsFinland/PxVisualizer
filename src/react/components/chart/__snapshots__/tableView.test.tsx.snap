// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableView render tests Should render correctly 1`] = `
<DocumentFragment>
  <div
    class="tableChart"
    id="foobar"
  >
    <table
      tabindex="0"
    >
      <caption
        class="tableChart-caption"
      />
      <thead>
        <tr>
          <td
            colspan="3"
            rowspan="2"
          />
          <th
            colspan="2"
            scope="col"
          >
            Yksiöt
          </th>
          <th
            colspan="2"
            scope="col"
          >
            Kaksiot
          </th>
        </tr>
        <tr>
          <th
            colspan="1"
            scope="col"
          >
            Lukumäärä
          </th>
          <th
            colspan="1"
            scope="col"
          >
            Neliövuokra (eur/m2)
          </th>
          <th
            colspan="1"
            scope="col"
          >
            Lukumäärä
          </th>
          <th
            colspan="1"
            scope="col"
          >
            Neliövuokra (eur/m2)
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            rowspan="8"
            scope="row"
          >
            Helsinki
          </th>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q1
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            13 021
          </td>
          <td>
            26,64
          </td>
          <td>
            10 080
          </td>
          <td>
            20,22
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            3 810
          </td>
          <td>
            15,90
          </td>
          <td>
            6 176
          </td>
          <td>
            13,74
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q2
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            11 231
          </td>
          <td>
            26,82
          </td>
          <td>
            9 326
          </td>
          <td>
            20,45
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            3 552
          </td>
          <td>
            15,97
          </td>
          <td>
            5 749
          </td>
          <td>
            13,78
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q3
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            10 269
          </td>
          <td>
            26,93
          </td>
          <td>
            8 907
          </td>
          <td>
            20,59
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            3 249
          </td>
          <td>
            15,99
          </td>
          <td>
            5 280
          </td>
          <td>
            13,79
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q4
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            10 272
          </td>
          <td>
            26,93
          </td>
          <td>
            8 805
          </td>
          <td>
            20,66
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            3 323
          </td>
          <td>
            16,08
          </td>
          <td>
            5 265
          </td>
          <td>
            13,80
          </td>
        </tr>
        <tr>
          <th
            rowspan="8"
            scope="row"
          >
            Vantaa
          </th>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q1
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            3 043
          </td>
          <td>
            23,48
          </td>
          <td>
            5 261
          </td>
          <td>
            17,62
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            941
          </td>
          <td>
            16,13
          </td>
          <td>
            2 083
          </td>
          <td>
            13,96
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q2
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            2 679
          </td>
          <td>
            23,64
          </td>
          <td>
            5 017
          </td>
          <td>
            17,79
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            846
          </td>
          <td>
            16,05
          </td>
          <td>
            1 902
          </td>
          <td>
            13,91
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q3
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            2 533
          </td>
          <td>
            23,68
          </td>
          <td>
            4 925
          </td>
          <td>
            17,85
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            763
          </td>
          <td>
            15,97
          </td>
          <td>
            1 778
          </td>
          <td>
            13,90
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q4
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            2 548
          </td>
          <td>
            23,77
          </td>
          <td>
            4 882
          </td>
          <td>
            17,99
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            767
          </td>
          <td>
            15,98
          </td>
          <td>
            1 744
          </td>
          <td>
            13,97
          </td>
        </tr>
        <tr>
          <th
            rowspan="8"
            scope="row"
          >
            Turku
          </th>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q1
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            5 534
          </td>
          <td>
            18,16
          </td>
          <td>
            4 267
          </td>
          <td>
            14,07
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            1 292
          </td>
          <td>
            13,91
          </td>
          <td>
            1 782
          </td>
          <td>
            11,81
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q2
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            4 437
          </td>
          <td>
            18,32
          </td>
          <td>
            3 666
          </td>
          <td>
            14,27
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            1 149
          </td>
          <td>
            14,02
          </td>
          <td>
            1 642
          </td>
          <td>
            11,87
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q3
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            4 014
          </td>
          <td>
            18,44
          </td>
          <td>
            3 329
          </td>
          <td>
            14,37
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            1 107
          </td>
          <td>
            14,06
          </td>
          <td>
            1 554
          </td>
          <td>
            11,86
          </td>
        </tr>
        <tr>
          <th
            rowspan="2"
            scope="row"
          >
            2022Q4
          </th>
          <th
            rowspan="1"
            scope="row"
          >
            Vapaarahoitteinen
          </th>
          <td>
            4 195
          </td>
          <td>
            18,53
          </td>
          <td>
            3 414
          </td>
          <td>
            14,48
          </td>
        </tr>
        <tr>
          <th
            rowspan="1"
            scope="row"
          >
            ARA
          </th>
          <td>
            1 161
          </td>
          <td>
            14,11
          </td>
          <td>
            1 557
          </td>
          <td>
            11,96
          </td>
        </tr>
      </tbody>
    </table>
    <p>
      Yksikkö: Lukumäärä: lukumäärä, Neliövuokra (eur/m2): eur / m2
    </p>
    <p>
      Lähde: PxVisualizer-fi
    </p>
  </div>
</DocumentFragment>
`;
